### PR TITLE
feature: support passing hyperparams directly to entry point function

### PR DESCRIPTION
**Pull request by ajberdy**
*Issue #, if available:*

Per @bpdoyle's feature suggestion

*Description of changes:*

Allow customers to add arguments to their entry point function to access hyperparameters. This is a non-breaking change, since previously, positional arguments were unsupported and kwargs ignored. Customers can still access hyperparameters through the environment variable.

```
# orchestration script

hyperparameters = {
    string_arg: "string",
    float_arg: 3.14,
}
```
```
# algorithm script (previously)

def entry_point():
    hp_file = os.environ["AMZN_BRAKET_HP_FILE"]
    with open(hp_file) as f:
        hyperparameters = json.load(f)
    string_arg = hyperparameters["string_arg"]
    float_arg = hyperparameters["float_arg"]
    print(string_arg)
    print(float_arg)
```
```
# algorithm script (now)

def entry_point(string_arg, float_arg=0):
    print(string_arg)
    print(float_arg)
```
```
# algorithm script (now, alternative)

def entry_point(**hyperparams):
    print(hyperparams["string_arg"])
    print(hyperparams["float_arg"])
```

*Testing done:*

Manual verification for all of the positional arguments, *args, and **kwargs in local/aws quantum jobs. Integ tests to be added to the bdk upon building the service containers with this change.

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [X] I have read the [CONTRIBUTING](https://github.com/aws/amazon-braket-containers/blob/main/CONTRIBUTING.md) doc
- [X] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/amazon-braket-containers/blob/main/CONTRIBUTING.md#commit-your-change)
- [X] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-braket-containers/blob/main/README.md) and [API docs](https://github.com/aws/amazon-braket-containers/blob/main/CONTRIBUTING.md#documentation-guidelines) (if appropriate)

#### Tests

- [X] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [X] I have checked that my tests are not configured for a specific region or account (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/braket_container.py [hunk 1]
```diff
@@ -149,14 +150,27 @@ def kick_off_customer_script(entry_point : str) -> multiprocessing.Process:
         str_module, _, str_method = entry_point.partition(":")
         customer_module = importlib.import_module(str_module)
         customer_method = getattr(customer_module, str_method)
-        customer_code_process = multiprocessing.Process(target=customer_method)
+
+        process_kwargs = {"target": customer_method}
+
+        hp_file = os.environ["AMZN_BRAKET_HP_FILE"]
```
Comment (krneta) [inline src/braket_container.py:156]: Is this the same as SM_HPS ?

File: src/braket_container.py [hunk 2]
```diff
@@ -149,14 +150,27 @@ def kick_off_customer_script(entry_point : str) -> multiprocessing.Process:
...
+        with open(hp_file) as f:
+            hyperparams = json.load(f)
```
Comment (krneta) [inline src/braket_container.py:158]: Hyperparameters can also be sent in env variables, do we need to handle both?

File: src/braket_container.py [hunk 3]
```diff
@@ -149,14 +150,27 @@ def kick_off_customer_script(entry_point : str) -> multiprocessing.Process:
...
+
+        try:
+            inspect.signature(customer_method).bind(**hyperparams)
```
Comment (krneta) [inline src/braket_container.py:161]: I thought that hyperparamters were converted to string by SageMaker, so you may have a float parameter, but you actually have to convert it to float, no?

**Comment by ajberdy** [comment]
tests failing around mocking of hp file, working on a fix

File: src/braket_container.py (hunk 1 above)
Comment (ajberdy) [reply src/braket_container.py:156]: It looks like it's similar, but SM_HPS is a string. Is that a stable API we can rely on? I believe AMZN_BRAKET_HP_FILE is the documented way of getting hyperparameters for us, but SM_HPS may be slightly more efficient

Comment (ajberdy) [reply src/braket_container.py:156]: Doc link: https://docs.aws.amazon.com/braket/latest/developerguide/braket-jobs-inputs-and-outputs.html#braket-jobs-environmental-variables

File: src/braket_container.py (hunk 2 above)
Comment (ajberdy) [reply src/braket_container.py:158]: I believe all user set env variables should be in the AMZN_BRAKET_HP_FILE since this is after the job goes through the service and the customer code is about to be run

File: src/braket_container.py (hunk 3 above)
Comment (krneta) [reply src/braket_container.py:161]: unit test?

Can you test the boundary cases (function has no parms, too many params, etc.)?

File: README.md [hunk 4]
```diff
@@ -8,9 +8,9 @@ Create an ECR repository in your AWS account. In this example, we'll assume it's
 
 1. Clone the repo and set the following environment variables:
     ```shell script
-    export ACCOUNT_ID=<YOUR_ACCOUNT_ID>
     export REGION=us-west-2
     export REPOSITORY_NAME=braket-base-jobs
+    export ACCOUNT_ID=<YOUR_ACCOUNT_ID>
```
Comment (ajberdy) [inline README.md:13]: I made this change so it's easier to copy and paste and run

File: src/braket_container.py [hunk 5]
```diff
@@ -149,14 +150,39 @@ def kick_off_customer_script(entry_point : str) -> multiprocessing.Process:
...
         customer_code_process.start()
     except Exception as e:
         log_failure_and_exit(f"Unable to run job at entry point {entry_point}\nException: {e}")
     return customer_code_process
 
 
-def join_customer_script(customer_code_process : multiprocessing.Process):
+def try_bind_hyperparameters_to_customer_method(customer_method: Callable):
+    hp_file = os.getenv("AMZN_BRAKET_HP_FILE")
+    with open(hp_file) as f:
```
Comment (krneta) [inline src/braket_container.py:169]: Should we check if the file isn't actually specified? Or what was the reason that you did not put this part inside the try statement?

File: README.md (hunk 4 above)
Comment (krneta) [reply README.md:13]: For my info, how does it make it easier?

Comment (ajberdy) [reply README.md:13]: If you copy and paste from the README into your terminal, it will automatically run the first two lines (because of the newlines) and paste the last into the terminal ready to go. This way, if you're using the default region/repo name, you can just change the last command. Otherwise, you have to copy and paste the commands in batches or find some other work around. Even if it was all on one line, it's easier to replace the suffix of a string rather than scrolling earlier. It's a small change, but a nice QOL improvement

File: src/braket_container.py (hunk 5 above)
Comment (ajberdy) [reply src/braket_container.py:169]: This is a stable part of the jobs API, whether or not hyperparameters are given (just confirmed with local/aws jobs). The try/except block exists very specifically to catch a TypeError on the `bind` method when hyperparameters do not line up with function parameters.

File: src/braket_container.py [hunk 6]
```diff
@@ -149,14 +150,39 @@ def kick_off_customer_script(entry_point : str) -> multiprocessing.Process:
...
+        inspect.signature(customer_method).bind(**hyperparameters)
+        annotations = inspect.getfullargspec(customer_method).annotations
+        function_args = {}
+        for param in hyperparameters:
+            function_args[param] = annotations.get(param, str)(
+                hyperparameters[param]
+            )
+        return function_args
+    except TypeError:
+        pass
```
Comment (ajberdy) [inline src/braket_container.py:182]: we could move things around here to isolate the try/except logic

```suggestion
    try:
        can_bind = inspect.signature(customer_method).bind(**hyperparameters)
    except TypeError:
        pass
    if can_bind:
        annotations = inspect.getfullargspec(customer_method).annotations
        function_args = {}
        for param in hyperparameters:
            function_args[param] = annotations.get(param, str)(
                hyperparameters[param]
            )
        return function_args
```

File: README.md (hunk 4 above)
Comment (krneta) [reply README.md:13]: Good call! I didn't think of that.

File: src/braket_container.py (hunk 5 above)
Comment (ajberdy) [reply src/braket_container.py:169]: Cleaned up the code to make this more clear

Comment (krneta) [reply src/braket_container.py:169]: I'm worried about SM changing it behind the scenes where if there are no HP then the file will not be given/created. In other parts of the code we protect against this, should we do the same here?

Comment (ajberdy) [reply src/braket_container.py:169]: To me, it seems that removing this file would be a breaking change. For example, a customer could have code:
```
# hp defaults
hps = {
    "hp_1": 1,
    "hp_2": 2,
    ...
}
provided_hps = load_hps_from_file()
hps.update(provided_hps)
```
and expect this to work whether or not hyperparameters are provided for that job run.

For that reason, it feels like we can assume that we will keep this contract and not get rid of that hp file. I believe that is something we own as a service and not created by sagemaker. There are a couple gaps in my knowledge of how exactly this file gets created, so happy to discuss further

Comment (krneta) [reply src/braket_container.py:169]: Yes, but I think that's a contrived example. Most customers that don't use hyperparameters wouldn't have code to handle them.

Comment (ajberdy) [reply src/braket_container.py:172]: Either way, doesn't hurt to have the check. Added

**Review by krneta — APPROVED** [review]
